### ISSUE TITLE
RL methods: sac.hpp is an implementation of TD3, not SAC.

### ISSUE THREAD

**Issue by gunnxx** [mlpack/mlpack#2923]
Upon reading the [`sac_impl.hpp`](https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/reinforcement_learning/sac_impl.hpp), I realized that it's not an implementation of Soft Actor Critic [SAC] ([Haarnoja et al., 2018](https://arxiv.org/abs/1801.01290)). The [`mlpack::rl::SAC::Update()`](https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/reinforcement_learning/sac_impl.hpp#L158) actually implements off-policy actor-critic with clipped double Q-learning, which is the key part of Twin Delayed DDPG [TD3] ([Fujimoto et al., 2018](https://arxiv.org/abs/1802.09477)). Although SAC makes use of clipped double Q-learning, the main component of SAC is augmenting standard RL objective with expected entropy of the policy (maximum entropy reinforcement learning). It can be clearly seen by comparing the [`mlpack::rl::SAC::Update()`](https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/reinforcement_learning/sac_impl.hpp#L158) method with [TD3 pseudocode](https://spinningup.openai.com/en/latest/algorithms/td3.html#pseudocode) as opposed to [SAC pseudocode](https://spinningup.openai.com/en/latest/algorithms/sac.html#pseudocode).

**Comment by medias661** [comment]
When reading it for a while I see some resemblance. If I trace back this implementation goes way back to GSOC 2020 I think @zoq  must have a better explanation about this.

**Comment by gunnxx** [comment]
@medias661 as far as my understanding, the resemblance comes from clipped double Q-learning, which was a key component proposed in TD3 paper. SAC algorithm makes use of clipped double Q-learning, however key part of SAC algorithm is the entropy maximization which is not present in the current implementation. Maybe I misunderstood, better to wait for explanation.

My GSoC project proposal is to add other reinforcement learning methods. Since each methods is disentangled in its implementation, I don't think it will affect my proposed algorithms. I stumbled upon this when I tried to read the whole RL methods code base (just to get an idea how my proposed algorithms will fit into existing code base).
